include/nlohmann/detail/iterators/iter_impl.hpp

```diff
@@ -463,11 +463,17 @@ class iter_impl // NOLINT(cppcoreguidelines-special-member-functions,hicpp-speci
 
     /*!
     @brief comparison: equal
-    @pre The iterator is initialized; i.e. `m_object != nullptr`.
+    @pre (1) The iterator is initialized; i.e. `m_object != nullptr`, or (2) both iterators are value-initialized.
     */
     template < typename IterImpl, detail::enable_if_t < (std::is_same<IterImpl, iter_impl>::value || std::is_same<IterImpl, other_iter_impl>::value), std::nullptr_t > = nullptr >
     bool operator==(const IterImpl& other) const
     {
+        // value-initialized forward iterators can be compared, and must compare equal to other value-initialized iterators of the same type #4493
```

Comment:
This can replace `JSON_ASSERT(m_object != nullptr);`.

Reply:
I would like to leave the assert in, because right below we dereference `m_object`.

Reply:
The dereferencing is already guarded by the two checks. The `JSON_ASSERT(m_object != nullptr);` assert is now completely redundant. The early exits due to `m_object != other.m_object` check and the new check `m_object == nullptr && other.m_object == nullptr` automatically imply  will always result in `m_object != nullptr` at the  assertion point.

Reply:
Yes, I understand. But then at some point, someone may see `m_object->m_data.m_type` and ask if this is safe given that `m_object` might be null. Then the assert above is even more explicit than adding a comment for which logical reason it cannot be.

So I see the following alternatives:

- Remove the assertion.
- Remove the assertion and add a note why the `m_object->m_data.m_type` is safe here.
- Leave the assertion in.

I am worried about coming back to this function after C++37 introduces some subtlety to iterators and ask the same questions.

(But maybe I'm just paranoid.)

Reply:
Ah, I see your confusion.  I'm not saying add a comment.  I'm saying do this:

```
    {
        // if objects are not the same, the comparison is undefined
        if (JSON_HEDLEY_UNLIKELY(m_object != other.m_object))
        {
            JSON_THROW(invalid_iterator::create(212, "cannot compare iterators of different containers", m_object));
        }

        // value-initialized forward iterators can be compared, and must compare equal to other value-initialized iterators of the same type #4493
        if (m_object == nullptr)
        {
            return true;
        }

        switch (m_object->m_data.m_type)
        {
```

Reply:
💡

Reply:
I did not get that you proposed to change the order of the checks - I just thought you wanted to remove the redundant assertion.

The code looks much better like this. Thanks for the patience!

Reply:
You're welcome, need to do the same for the `operator<` though.